subprojects/core/src/main/java/org/gradle/api/tasks/options/Option.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.tasks.options;
+
+import org.gradle.api.Incubating;
+
+import java.lang.annotation.*;
+
+/**
+ * <p>Marks a property of a {@link org.gradle.api.Task} as being configurable from the command-line.</p>
+ *
+ * <p>This annotation should be attached to a field or a setter method. When attached to a field, {@link #option()}
+ * will use the name of the field by default. When attached to a method, {@link #option()} must be specified.</p>
+ *
+ * <p>An option may have one of the following types:</p>
+ * <ul>
+ * <li>{@code boolean}</li>
+ * <li>{@code Boolean}</li>
+ * <li>{@code Enum}</li>
+ * <li>{@code List&lt;Enum&gt;}</li>
+ * <li>{@code List&lt;String&gt;}</li>
+ * <li>{@code String}</li>
+ * </ul>
+ *
+ * @since 4.6
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD, ElementType.FIELD})
+@Inherited
+@Incubating
+public @interface Option {
+
+    /**
+     * The option to map to this property. Required when annotating a method. May be omitted when annotating a field
+     * in which case the field's name will be used.
+     *
+     * @return The option.
+     */
+    String option() default "";
+
+    /**
+     * The description of this option.
+     *
+     * @return The description.
+     */
+    String description();
+
+    /**
+     * The order of this option for displaying in help command output. Allows overriding default alphabetical order.
+     *
+     * @return The order.
+     */
+    int order() default 0;
```

Comment:
⭕️ Do we have a use case for this?

Reply:
I think the original intention was to be able to order the options in the output of the `help` task independent of alphabetic order (which is the default). At the moment this is used by the `InitBuild` task type.

Looking at the history of the changes it seems that @pioterj introduced this parameter. Can you tell us more about the intention?

Reply:
If we don't use it, we shouldn't make it part of the public API.

Reply:
We do use it.

Reply:
I meant if we don't have a strong reason for it. If we support it we have to make sure it doesn't explode when used with subclasses adding more properties etc.

Reply:
Wow, that was a while ago. @bmuschko It was introduced in this commit: https://github.com/gradle/gradle/commit/d808873dd975469d624314d8ef61990bb63fc57d. If I recall correctly, I added it to display `--type` option before `--test-framework` in build init help output because the first one is the one that is required and more important. It looked weird with alphabetical ordering where `--test-framework` was first. Now that I think about it, perhaps a better way to model it would be to provide a way to describe which option is required and display required options first in help output.

Reply:
Thanks for the explanation. At the moment we have no parameter for indicating that an option is mandatory or optional. That's something we can add later. Even an ordering of option doesn't clearly indicate that. I may just remove that parameter altogether and then we think about a better way to model it in the future.